### PR TITLE
Integrate with submodules

### PR DESCRIPTION
**Pull request by ten3roberts**
This enables neogit to work with submodules

image

Fixes: #870

### CODE REVIEW

File: lua/neogit/status.lua [hunk 1]
```diff
@@ -289,18 +306,26 @@ local function draw_buffer()
...
         elseif f.mode then
-          output:append(string.format("%s %s", label, f.name))
+          line = string.format("%s %s", label, f.name)
         else
-          output:append(f.name)
+          line = f.name
+        end
+
+        if f.submodule then
+          line = line .. " " .. format_submodule_mode(f.submodule)
```
Comment (CKolkey) [inline lua/neogit/status.lua:324]: Definitely `string.format` here. Lua performance can go to shit with a lot of string concatenation.

**Comment by ten3roberts** [comment]
In order to get submodules working properly I found it necessary to make the status buffer not be a global state but rather instance based. The reason is that opening and going back and forth between submodules status panes would make these status buffers trample that same global state.

This approach makes the status buffer an object which can be manipulated and returned/found based on the git repositories root, and will in theory allow having multiple instances of neogit for different repositories open at the same time in tabs, splits, or the background.

**Comment by CKolkey** [comment]
Cool, the change makes sense. I always thought the global status buffer thing was weird conceptually, but :shrug: it's from before my time here.

I'd say don't worry too much about the status buffer - I'm going to be replacing it soon with a rewrite built using the UI framework. I finally got a couple building blocks in place to make that work:
- `on_open` callback function for "folded" content for lazy-loading diffs
- an index built into the renderer so finding components under the cursor is (nearly) `o(1)`. All components are indexed by line number in a table, so finding the right one is much faster than iterating through all components starting at line 1.

Still a WIP... lots of functions to pull in from the existing buffer. If you're up for it I wouldn't mind help :) But all good either way.

https://github.com/NeogitOrg/neogit/pull/1161

**Comment by ten3roberts** [comment]
Awesome!

Thanks for the heads up

I think both of these rewrites would go quite well together.

Will daily run this one for a bit and see how it fairs and fix any issues (and get the tests to pass).

Have some more time and mental energy now to pursue my side projects, which feels fantastic :smile: 

Don't worry too much about merge conflicts, as the Swedish saying goes: "As one makes their bed, so they must lie"

**Comment by CKolkey** [comment]
Given the number of conflicts, I think it's safe to assume this can be closed. However, I've copied the core change - displaying submodule status - here: 51450329e04860f71a31d8c7ef5178fce6e60524

**Pull request closed without merging**